edk2toollib/uefi/telem/parser_main.py

```diff
@@ -0,0 +1,660 @@
+FRU ID              (32 byte offset, 16 byte length) : ? Not detailed in the CPER doc
+Section Severity    (48 byte offset, 4  byte length) : Value from 0 - 3 indicating the severity of the error
+FRU String          (52 byte offset, 20 byte length) : String identifying the FRU hardware
+
+
+Python Struct Format Characters
+
+Format      C Type                  Python Type     Standard Size
+x           pad byte                none            1
+c           char                    integer         1
+b           signed char             integer         1
+B           unsigned char           integer         1
+?           _Bool                   bool            1
+h           short                   integer         2
+H           unsigned short          integer         2
+i           int                     integer         4
+I           unsigned int            integer         4
+l           long                    integer         4
+L           unsigned long           integer         4
+q           long long               integer         8
+Q           unsigned long long      integer         8
+n           ssize_t                 integer
+N           size_t                  integer
+e           (6)                     float           2
+f           float                   float           4
+d           double                  float           8
+s           char[]                  bytes
+p           char[]                  bytes
+P           void *                  integer
+"""
+
+CPER_HEADER_SIZE            = 128 # A CPER header is 128 bytes 
+CPER_SECTION_HEADER_SIZE    = 72  # A CPER section header is 72 bytes
+FriendlyNames               = {}  # dict loaded from friendlynames.csv used to associate guids with friendly strings
+Parsers                     = []  # list of plugin classes which inherit from CPER_SECTION_DATA and are therefore capable of parsing section data
+
+# Using these lambda concat statements to guarantee linear runtime when concatonating strings, and for simplicity
+concat_s = lambda x,y  : (''.join([x,y," "]))  # Concatenate two strings and put a space at the end
+concat_n = lambda x,y  : (''.join([x,y,"\n"])) # Concatenate two strings and put a newline at the end
+concat   = lambda x,y  : (''.join([x,y]))      # Concatenate two strings
+
+# Useful lambdas
+empty    = lambda x     : x == "" or x == None or x == []   # True if x is an empty string, list, etc.
+neq      = lambda x,y   : x != y                            # True if x and y are not equal
+eq       = lambda x,y   : x == y                            # True if x and y are equal
+
+class CPER(object):
+
+    def __init__(self, input):
+        self.RawData = bytearray.fromhex(input)
+        self.Header = None
+        self.SetCPERHeader()
+        self.SectionHeaders = []
+        self.SetSectionHeaders()
+
+    ##
+    # Turn the portion of the raw input associated with the CPER head into a CPER_HEAD object
+    ##
+    def SetCPERHeader(self):
```

Comment:
Should this be a private function? It looks like we only do it in the beginning. _SetCPERHeader